File: workflow/scripts/main_figure_1_and_2/downsampling_reads_for_UMIs.py
# Script: downsampling_reads_for_UMIs.py

### SETUP =====================================================================

# Saving state for debugging
import sys
import pickle

def save_state(filename):
	with open(f"RDA_objects/downsampling_reads_for_UMIs.pkl", "wb") as f:
		pickle.dump(globals(), f, protocol=pickle.HIGHEST_PROTOCOL)
	print("Saved state")

save_state("downsampling_reads_for_UMIs")
# sys.exit()  # Uncomment to stop manually after saving state

# Open log file for messages, warnings, and errors
log = open(snakemake.log[0], "w")
sys.stdout = log
sys.stderr = log


### LOADING FILES =============================================================

print("Loading packages")
import numpy as np
import pandas as pd
import h5py
import matplotlib.pyplot as plt
import seaborn as sns
from itertools import product
import scanpy as sc
import os

print("Loading input files")
# Get dataset paths from Snakemake input object
datasets = [
	{
		"name": "DC-TAP-seq (MOI6)",
		"mol_info": snakemake.input.moi6_mol_info,
		"filtered_h5": snakemake.input.moi6_filtered_h5
	},
	{
		"name": "CRISPRi Direct Capture",
		"mol_info": snakemake.input.crisprdi_mol_info,
		"filtered_h5": snakemake.input.crisprdi_filtered_h5
	},
	{
		"name": "DC-TAP-seq (MOI3)",
		"mol_info": snakemake.input.moi3_mol_info,
		"filtered_h5": snakemake.input.moi3_filtered_h5
	}
]

print(f"Will process {len(datasets)} datasets:")
for dataset in datasets:
	print(f"  - {dataset['name']}")
	

### ANALYSIS =========================================================

def get_complexity_df(read_df, dataset_name, fracs=None):
	"""
	Process reads data to perform downsampling analysis.
	This follows Maddie's approach but without the TPM related parts.
	"""
	# Get the raw reads
	read_df['reads'] = read_df['count'].astype(int)
	read_df.rename(columns={'barcode_idx': 'cell'}, inplace=True)
	
	# Create a dataframe with each read represented by a row
	print(f"Expanding reads for {dataset_name}...")
	df_repeat = read_df.reindex(read_df[['umi', 'reads']].index.repeat(read_df.reads))
	
	# Define downsampling fractions (0% to 100%)
	if fracs is None:
		fracs = np.arange(0.1, 1.1, 0.1)
	
	# Create empty results dataframe
	results = pd.DataFrame(index=range(len(fracs)),
						  columns=['proportion', 'mean_reads_per_cell', 'mean_umis_per_cell_per_gene', 'dataset'])
	
	# Downsample at each fraction
	for i, frac in enumerate(fracs):
		print(f"Processing fraction {frac} for {dataset_name}")
		
		# Take a sampling of our reads at that fraction
		sample = df_repeat.sample(frac=frac, random_state=42)
		
		# Total reads leftover after sampling
		total_reads = sample.shape[0]
		
		# Drop the duplicate reads now to count unique molecular identifiers per cell per gene
		umis_cell_gene = sample.drop_duplicates().groupby(['cell', 'gene_name']).count()['umi'].reset_index()
		
		# Remake a cell counts matrix (cell x gene with umi counts per gene in that cell)
		cell_counts_matrix = umis_cell_gene.pivot(index='cell', columns='gene_name').fillna(0)
		
		# Get metrics from that matrix
		mean_umis_per_cell_per_gene = cell_counts_matrix.mean().mean()
		
		# Total reads per cell considers all in the sampling
		mean_reads_per_cell = total_reads / len(sample.cell.unique())
		
		# Store results
		results.loc[i, 'proportion'] = frac
		results.loc[i, 'mean_reads_per_cell'] = mean_reads_per_cell
		results.loc[i, 'mean_umis_per_cell_per_gene'] = mean_umis_per_cell_per_gene
		results.loc[i, 'dataset'] = dataset_name
	
	return results

def process_dataset(dataset):
	"""Process a single dataset using Maddie's approach"""
	name = dataset["name"]
	mol_info_path = dataset["mol_info"]
	filtered_h5_path = dataset["filtered_h5"]
	
	print(f"Processing {name}...")
	
	# Check if files exist
	if not os.path.exists(mol_info_path) or not os.path.exists(filtered_h5_path):
		print(f"Files not found for dataset: {name}")
		return None
	
	try:
		# Read in final output of counts
		counts = sc.read_10x_h5(filtered_h5_path)
		ensg_id_map = counts.var.reset_index()
		
		# Read in molecule_info.h5
		mol_dfs = {}
		names = 'barcode_idx umi count feature_idx'.split()
		for name_field in names:
			with h5py.File(mol_info_path, 'r') as f:
				mol_dfs[name_field] = pd.DataFrame(np.array(f[name_field]))
		
		molecule_info = pd.concat(mol_dfs, axis=1)
		molecule_info.columns = molecule_info.columns.droplevel(1)
		
		# Read features and barcodes
		with h5py.File(mol_info_path, "r") as f:
			ds_obj = f['features']
			gene_id_df = pd.DataFrame(ds_obj['id'][()].astype(str))
			
		with h5py.File(mol_info_path, "r") as f:
			ds_obj = f['barcodes']
			barcode_id_df = pd.DataFrame(ds_obj[()].astype(str))
		
		# Get the actual cell names and gene names
		molecule_info['barcode_idx'] = molecule_info['barcode_idx'].map(lambda x: barcode_id_df.iloc[x, 0] if x < len(barcode_id_df) else None)
		molecule_info['feature_idx'] = molecule_info['feature_idx'].map(lambda x: gene_id_df.iloc[x, 0] if x < len(gene_id_df) else None)
		
		# Map to gene names from the counts matrix
		if isinstance(counts.var.index, pd.MultiIndex):
			gene_map = dict(zip(counts.var.index.get_level_values(0), counts.var.index.get_level_values(1)))
		else:
			gene_map = dict(zip(ensg_id_map['gene_ids'], ensg_id_map['index'] if 'index' in ensg_id_map.columns else ensg_id_map.index))
			
		molecule_info['gene_name'] = molecule_info['feature_idx'].map(gene_map)
		molecule_info.dropna(subset=['gene_name'], inplace=True)
		
		# Filter to cells in the counts matrix
		cell_barcodes = counts.obs.index.str.replace('-1$', '', regex=True)
		reads_df = molecule_info[molecule_info['barcode_idx'].isin(cell_barcodes)]
		
		# Run the downsampling analysis
		if not reads_df.empty:
			return get_complexity_df(reads_df, name)
		else:
			print(f"No matching cells found for {name}")
			return None
			
	except Exception as e:
		print(f"Error processing {name}: {str(e)}")
		return None

# Process all datasets
all_results = []

for dataset in datasets:
	results = process_dataset(dataset)
	if results is not None:
		all_results.append(results)
		

### SAVE OUTPUT ===============================================================

print("Saving output files")

# Combine all results
if all_results:
	combined_results = pd.concat(all_results)
	
	# Save the raw results
	combined_results.to_csv(snakemake.output.combined_results, index=False)
	
	# Create plot
	plt.figure(figsize=(10, 6))
	sns.lineplot(data=combined_results, x='mean_reads_per_cell', y='mean_umis_per_cell_per_gene', 
				 hue='dataset', marker='o', markersize=8)
	
	plt.title('DC-TAP-seq Effect Size Analysis', fontsize=16, fontweight='bold')
	plt.xlabel('Mean Total Reads per Cell', fontsize=14)
	plt.ylabel('Mean UMIs per Cell per Gene', fontsize=14)
	plt.grid(alpha=0.3)
	plt.tight_layout()
	
	# Set colors similar to the R plot
	colors = ["#4477AA", "#EE6677", "#228833"]
	for i, line in enumerate(plt.gca().get_lines()):
		if i < len(colors):
			line.set_color(colors[i])
	
	plt.savefig(snakemake.output.plot_pdf)
	plt.savefig(snakemake.output.plot_png)
	print("Analysis complete. Plots saved")
else:
	print("No data to plot. Check file paths and processing.")
	

### CLEAN UP ==================================================================

print("Closing log file")
log.close()

File: workflow/scripts/main_figure_1_and_2/downsampling_reads_for_UMIs.R
# Script: downsampling_reads_for_UMIs.R

### SETUP =====================================================================

# Saving image for debugging
save.image(paste0("RDA_objects/downsampling_reads_for_UMIs.rda"))
message("Saved Image")
# stop("Manually Stopped Program after Saving Image")

# Open log file to collect messages, warnings, and errors
log_filename <- snakemake@log[[1]]
log <- file(log_filename, open = "wt")
sink(log)
sink(log, type = "message")


### LOADING FILES =============================================================

message("Loading packages")
suppressPackageStartupMessages({
  library(tidyverse)
  library(rhdf5)
  library(Seurat)
})

message("Loading input files")
# Get dataset paths from Snakemake input object
datasets <- list(
  list(
    name = "DC-TAP-seq (MOI6)",
    mol_info = snakemake@input$moi6_mol_info,
    filtered_h5 = snakemake@input$moi6_filtered_h5
  ),
  list(
    name = "CRISPRi Direct Capture",
    mol_info = snakemake@input$crisprdi_mol_info,
    filtered_h5 = snakemake@input$crisprdi_filtered_h5
  ),
  list(
    name = "DC-TAP-seq (MOI3)",
    mol_info = snakemake@input$moi3_mol_info,
    filtered_h5 = snakemake@input$moi3_filtered_h5
  )
)

message(paste("Will process", length(datasets), "datasets:"))
for(dataset in datasets) {
  message(paste("  -", dataset$name))
}


### ANALYSIS FUNCTIONS ========================================================

get_complexity_df <- function(read_df, dataset_name, fracs = NULL) {
  message(paste("Processing reads data for", dataset_name))
  
  # Get the raw reads
  read_df$reads <- as.integer(read_df$count)
  read_df <- read_df %>% rename(cell = barcode_idx)
  
  # Create a dataframe with each read represented by a row
  message(paste("Expanding reads for", dataset_name, "..."))
  df_repeat <- read_df[rep(row.names(read_df), read_df$reads), c("umi", "cell", "gene_name")]
  
  # Define downsampling fractions (10% to 100%)
  if(is.null(fracs)) {
    fracs <- seq(0.1, 1.0, 0.1)
  }
  
  # Create empty results dataframe
  results <- data.frame(
    proportion = numeric(length(fracs)),
    mean_reads_per_cell = numeric(length(fracs)),
    mean_umis_per_cell_per_gene = numeric(length(fracs)),
    dataset = character(length(fracs)),
    stringsAsFactors = FALSE
  )
  
  # Downsample at each fraction
  for(i in seq_along(fracs)) {
    frac <- fracs[i]
    message(paste("Processing fraction", frac, "for", dataset_name))
    
    # Take a sampling of our reads at that fraction
    set.seed(42)
    sample_indices <- sample(nrow(df_repeat), size = floor(nrow(df_repeat) * frac))
    sample_df <- df_repeat[sample_indices, ]
    
    # Total reads leftover after sampling
    total_reads <- nrow(sample_df)
    
    # Drop the duplicate reads now to count unique molecular identifiers per cell per gene
    umis_cell_gene <- sample_df %>%
      distinct() %>%
      group_by(cell, gene_name) %>%
      summarise(umi_count = n(), .groups = "drop")
    
    # Remake a cell counts matrix (cell x gene with umi counts per gene in that cell)
    cell_counts_matrix <- umis_cell_gene %>%
      pivot_wider(names_from = gene_name, values_from = umi_count, values_fill = 0)
    
    # Get metrics from that matrix
    mean_umis_per_cell_per_gene <- cell_counts_matrix %>%
      select(-cell) %>%
      summarise_all(mean) %>%
      unlist() %>%
      mean()
    
    # Total reads per cell considers all in the sampling
    mean_reads_per_cell <- total_reads / length(unique(sample_df$cell))
    
    # Store results
    results$proportion[i] <- frac
    results$mean_reads_per_cell[i] <- mean_reads_per_cell
    results$mean_umis_per_cell_per_gene[i] <- mean_umis_per_cell_per_gene
    results$dataset[i] <- dataset_name
  }
  
  return(results)
}

process_dataset <- function(dataset) {
  name <- dataset$name
  mol_info_path <- dataset$mol_info
  filtered_h5_path <- dataset$filtered_h5
  
  message(paste("Processing", name, "..."))
  
  # Read in final output of counts using Seurat
  counts <- Read10X_h5(filtered_h5_path)
  if(is.list(counts)) {
    counts <- counts$`Gene Expression`  # Take gene expression if multiple assays
  }
  
  # Get gene info - handle both list and matrix cases
  counts_no_names <- Read10X_h5(filtered_h5_path, use.names = FALSE)
  if(is.list(counts_no_names)) {
    gene_ids <- rownames(counts_no_names$`Gene Expression`)
  } else {
    gene_ids <- rownames(counts_no_names)
  }
  
  gene_info <- data.frame(
    gene_ids = gene_ids,
    gene_names = rownames(counts),
    stringsAsFactors = FALSE
  )
  
  # Read in molecule_info.h5
  message("Reading molecule info file...")
  
  # Read molecular info tables
  barcode_idx <- h5read(mol_info_path, "barcode_idx")
  umi <- h5read(mol_info_path, "umi")
  count <- h5read(mol_info_path, "count")
  feature_idx <- h5read(mol_info_path, "feature_idx")
  
  # Read barcodes and gene IDs
  barcodes <- h5read(mol_info_path, "barcodes")
  gene_ids <- h5read(mol_info_path, "features/id")
  
  h5closeAll()
  
  # Create molecule info dataframe
  molecule_info <- data.frame(
    barcode_idx = barcode_idx,
    umi = umi,
    count = count,
    feature_idx = feature_idx,
    stringsAsFactors = FALSE
  )
  
  # Map barcode indices to actual barcodes (vectorized - much faster!)
  barcode_lookup <- as.character(barcodes)
  idx_1based <- molecule_info$barcode_idx + 1  # Convert to 1-based indexing
  valid_mask <- idx_1based >= 1 & idx_1based <= length(barcode_lookup)
  molecule_info$barcode_idx <- NA_character_
  molecule_info$barcode_idx[valid_mask] <- barcode_lookup[idx_1based[valid_mask]]
  
  # Map feature indices to gene IDs (vectorized - much faster!)
  gene_lookup <- as.character(gene_ids)
  gene_idx_1based <- molecule_info$feature_idx + 1  # Convert to 1-based indexing
  gene_valid_mask <- gene_idx_1based >= 1 & gene_idx_1based <= length(gene_lookup)
  molecule_info$feature_idx <- NA_character_
  molecule_info$feature_idx[gene_valid_mask] <- gene_lookup[gene_idx_1based[gene_valid_mask]]
  
  # Map to gene names from the counts matrix
  gene_map <- setNames(gene_info$gene_names, gene_info$gene_ids)
  molecule_info$gene_name <- gene_map[molecule_info$feature_idx]
  
  # Filter to only the 93 amplified genes that are in your counts matrix
  # This removes the other ~35,000 genes you don't care about
  molecule_info <- molecule_info[!is.na(molecule_info$gene_name), ]
  
  # Filter to cells in the counts matrix
  cell_barcodes <- gsub("-1$", "", colnames(counts))
  reads_df <- molecule_info[molecule_info$barcode_idx %in% cell_barcodes, ]
  
  message(paste("Final dataset has", nrow(reads_df), "molecular observations for analysis"))
  
  # Run the downsampling analysis
  return(get_complexity_df(reads_df, name))
  
}


### MAIN ANALYSIS =============================================================

message("Processing all datasets")

# Process all datasets
all_results <- list()

for(i in seq_along(datasets)) {
  results <- process_dataset(datasets[[i]])
  if(!is.null(results)) {
    all_results[[i]] <- results
  }
}

# Remove NULL results
all_results <- all_results[!sapply(all_results, is.null)]


### CREATE PLOTS ==============================================================

message("Creating plots")

# Combine all results
combined_results <- do.call(rbind, all_results)

# Create plot
saturation_plot <- ggplot(combined_results, aes(x = mean_reads_per_cell, y = mean_umis_per_cell_per_gene, color = dataset)) +
  geom_line(linewidth = 1.2) +
  geom_point(size = 3) +
  scale_color_manual(values = c("#4477AA", "#EE6677", "#228833")) +
  labs(
    title = "DC-TAP-seq Effect Size Analysis",
    x = "Mean Total Reads per Cell",
    y = "Mean UMIs per Cell per Gene",
    color = "Dataset"
  ) +
  theme_classic() +
  theme(
    plot.title = element_text(size = 16, face = "bold"),
    axis.title = element_text(size = 14),
    axis.text = element_text(size = 12),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10),
    panel.grid.minor = element_blank()
  )

message("Analysis complete. Saving outputs...")
  

### SAVE OUTPUT ===============================================================

message("Saving output files")

# Save the raw results
write_csv(combined_results, "combined_results.tsv")
write_csv(combined_results, snakemake@output$combined_results)

# Save plots
ggsave(filename = snakemake@output$plot_pdf, plot = saturation_plot, width = 10, height = 6, device = "pdf")
saveRDS(saturation_plot, file = "saturation_plot.rds")


### CLEAN UP ==================================================================

message("Closing log file")
sink()
sink(type = "message")
close(log)
